proposals/0051-function-result-sigs.rst

```diff
@@ -242,8 +242,36 @@ Alternatives
   this is neither consistent nor terribly useful.
 
 * We could detect CUSKs as we do in types to enable polymorphic recursion, but
-  this makes little sense as we are in the proccess of their deprecation.
+  this makes little sense as we are in the process of their deprecation.
 
+* We could allow contexts in the return type.
+  An earlier version of this proposal had::
+
+    funlhs -> funlhs' [:: [context =>] type]
+
+  But the usage of this can be confusing.
+  For example::
+
+    f (x :: a) :: Eq a => a
+
+  looks like it it might have type::
+
+    f :: a -> Eq a => a
+
+  Which is gratuitously higher-ranked. Base Haskell 2010 never reorders
+  arguments currently, so just moving the ``Eq a`` to the front seems overly
+  magical.
+
+  However, with ``-XRank2Types``, contexts are allowed in types and so this
```

Comment:
I don't think we need the `RankNTypes` extension to enable contexts. It should just work.

Reply:
Without the `-XRank2Types`
```
Prelude> f :: Int -> Eq a => a; f = undefined

<interactive>:1:6: error:
    • Illegal qualified type: Eq a => a
      Perhaps you intended to use RankNTypes or Rank2Types
    • In the type signature: f :: Int -> Eq a => a
```

Reply:
That's true. The extension is slightly misnamed. Perhaps `FlexibleQuantification` would have been better. But I still think this new behavior can be enabled without the extension: the type you write at the end of the LHS will be prenex, and the inferred type of the function will be prenex (because of reshuffling the order), and so no one gets hurt.

Reply:
I...wish to be conservative here. It just occurred to me, say in some dependent future we have 
```haskell
f (a :: k) :: Constraint k a => f a = ...
```
This should infer to
```haskell
f :: forall k. foreach (a :: k) -> Constraint k a => f a
```
We can always be more flexible about rearranging things later, but once the cat is out of the bag we cannot put it back.

Reply:
The rearranging is all up to type inference -- it's not something that even needs to be specified here. Instead, we just say that the result has a certain type, and then type inference will either check a type signature against that or infers a type; either way, it all just works.

Reply:
@goldfirere I suppose with the normal "it just needs to unify" LHS treatment that is true.

So it sounds like you don't think anything should be changed in the proposal then? Just the wording cleaned up? We always allow contexts, but they and the return type signatures are optional?